protocol/proto/common.go

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package proto
+
+import (
+	"encoding/binary"
+
+	"github.com/m3db/m3x/pool"
+)
+
+const (
+	sizeBufferSize = 4
+)
+
+var sizeEncodeDecoder = binary.BigEndian
+
+func growDataBufferIfNeeded(buffer []byte, target int, pool pool.BytesPool) []byte {
+	if len(buffer) >= target {
+		return buffer
+	}
+	if pool != nil {
+		pool.Put(buffer)
+	}
+	return getByteSliceWithLength(target, pool)
+}
+
+func getByteSliceWithLength(target int, pool pool.BytesPool) []byte {
+	if pool == nil {
+		return make([]byte, target)
+	}
+	b := pool.Get(target)
+	// Make sure there is enough length in the slice.
+	return b[:cap(b)]
```

Comment:
it's better to do cap, for example the pool has {10, 512, 1024} buckets, and we did a Get(100), it will return a slice with 512 cap, but next time when we want to grow the slice to 200, we need to check the cap and reset the len, I feel it's simpler to just get the full cap in the first time.

Reply:
Hm in that case should you compare `cap(buffer)` against `targetSize` then?

Reply:
in this case len == cap, it's safer to compare len with targetSize no? in case the len is not at cap and the user may want to access idx that is between [len, cap)